tests/shard_map_test.py

```diff
@@ -230,6 +230,23 @@ def fwd(a):
     sum_of_odd_columns = np.sum(a[..., axis_index_groups[1]], -1)
     for i, sums in enumerate(np.split(sum_of_odd_columns, 4, 0)):
       self.assertAllClose(np.squeeze(c.addressable_data(2 * i + 1), -1), sums)
+  
+  def test_collective_broadcast(self):
```

Comment:
Can you test the broadcasting and autodiff rules too?

Reply:
Yes, let me add those.

Reply:
Added